metrics/prometheus/metrics.go

```diff
@@ -1,42 +1,120 @@
+// as per the Prometheus "best practices" in their documentation.
+// See: https://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics
+//
+// The labels associated with metrics should be exclusively descriptive.
+// For example, one might think that this is an applicable demonstration of "hydra_consent_requests_accepted":
+//   hydra_consent_requests_accepted{scope="scope:1"} 1
+//   hydra_consent_requests_accepted{scope="scope:2"} 1
+// when a consent request requests []string{"scope:1", "scope:2"}.
+// This is not good; if we were to query Prometheus for `sum(hydra_consent_requests_accepted)`, we would get 2, when in reality,
+// only 1 consent request was accepted. Instead, separate data like "scopes" into a more descriptive time series:
+//   hydra_consent_requests_accepted{scopes="2"} 1
+//   hydra_consent_requests_accepted_scopes{scope="scope:1"} 1
+//   hydra_consent_requests_accepted_scopes{scope="scope:2"} 1
+//
+// Some of these values will be populated by HTTP middleware, while others
+// will be populated elsewhere throughout the project
+type Metrics struct {
+	HTTPRequestCount      *prometheus.CounterVec
+	HTTPRequestDuration   *prometheus.HistogramVec
+	HTTPResponseSizeBytes *prometheus.HistogramVec
+
+	ConsentRequests         *prometheus.CounterVec
+	ConsentRequestScopes    *prometheus.CounterVec
+	ConsentRequestAudiences *prometheus.CounterVec
+	ConsentRequestsRejected *prometheus.CounterVec
+
+	LoginRequests        *prometheus.CounterVec
+	LoginRequestSubjects *prometheus.CounterVec
+
+	AccessTokensIssued   *prometheus.CounterVec
+	AccessTokensRevoked  *prometheus.CounterVec
+	RefreshTokensIssued  *prometheus.CounterVec
+	RefreshTokensRevoked *prometheus.CounterVec
+	IDTokensIssued       *prometheus.CounterVec
+
+	Grants  *prometheus.GaugeVec
+	Clients *prometheus.GaugeVec
+	Keys    *prometheus.GaugeVec
+}
+
+// NewMetrics registers our metrics and their labels to the prometheus Registry.
 func NewMetrics(version, hash, date string) *Metrics {
-	pm := &Metrics{}
+	pm := &Metrics{
+		HTTPRequestCount: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "http_request_count",
+			Help: "",
+		}, []string{"code", "method"}),
+		HTTPRequestDuration: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name:    "http_request_duration_seconds",
+				Help:    "A histogram of latencies for requests",
+				Buckets: []float64{.05, .25, .5, 1, 2.5, 5, 10},
+			},
+			[]string{"method"},
+		),
+		HTTPResponseSizeBytes: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name:    "http_response_size_bytes",
+				Help:    "A histogram of response sizes for requests",
+				Buckets: []float64{200, 400, 800, 1600},
+			},
+			[]string{},
+		),
+		ConsentRequests: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: "hydra",
+			Subsystem: "consent",
+			Name:      "requests",
+			Help:      "incremented when a request is sent to consent.AcceptConsentRequest",
```

Comment:
What about `reject`?

Reply:
I created a separate metric for `RejectConsentRequest` [here](https://github.com/ory/hydra/pull/1381/files/b0b492c3823b70296b90eca69d2872746595b60e#diff-01645a5b90df989c7ec7d047661882dfR90)